### ISSUE TITLE
a small bug in eulerian.eulerize ?

### ISSUE THREAD

**Issue by claudinepeyrat06** [networkx/networkx#5952]
I think that there is a small bug in eulerian.eulerize:
it needs to find a minimum weight perfect matching , and to do that use a maximum weight  matching algorithm replacing each weight w by its inverse 1/w, which is not the correct way It should  rather use matching.min_weight_matching , that correctly replace each weight w by C-w (where C is a constant  that make sure that the weight are positive)

### Steps to Reproduce

G= nx.cycle_graph(18)
G.add_edge(0,18)
G.add_edge(18,19)
G.add_edge(17,19)
G.add_edge(4,20)
G.add_edge(20,21)
G.add_edge(21,22)
G.add_edge(22,23)
G.add_edge(23,24)
G.add_edge(24,25)
G.add_edge(25,26)
G.add_edge(26,27)
G.add_edge(27,28)
G.add_edge(28,13)
eulerize(G) return a multigraph with 10 additional edges while it should have added only 8.
In this graph, odd degree vertices are :0,4,13,17. 
distance(0,17)=1  distance(4,13)=9
distance(0,4)=distance(13,17)= 4

eulerize chose to double 10 edges, while  it should have doubled only 8, Indeed 8<10 but unfortunaly  1/4+1/4 is also smaller that 1+1/9

**Comment by dschult** [comment]
I can verify that this is a bug!  And your solution should work well. I suggest that we choose C to be the maximum weight plus 1, though there might be a better choice. The ensures that integer weights stay integers to avoid round-off (another problem with using reciprocal edge weights). 

We need to create a PR with the fix and also a test using the example provided above to ensure that it is getting the correct answer. If you'd like to do that, it would be great (and faster) but we will get to it if you don't do that.

**Comment by claudinepeyrat06** [comment]
I will try ....

**Comment by claudinepeyrat06** [comment]
@dschult 
I think that eulerize() would be better if it were  slightly modified  :
- now it aims to add as few edges as possible to the graph to make it an eulerian multigraph, would it not be more general if it was aiming to minimize not the number of added  edges but the total weight of these edges ? 

For my purpose what i actually need is  to travel all edges at least once, at minimal cost,  but need to start and end at specified vertices,  I don't know if this would be  an interesting  option.

**Comment by dschult** [comment]
If I understand correctly, eulerian doesn't involve egde weights on the graph itself -- but this algorithm for "eulerizing" (is that even a word?) a graph tries to add as few edges as possible by creating an edge weight. But that edge weight is not related to the edge weights on the original graph.

Adding algorithms that deal with a weighted definition of "eulerian" sound like it is moving into new territory. The concept of eulerian wants all edges, so the total weight will be the same for every eulerian path. But now we are talking about eulerizing the graph with an optimality condition based on edge weights.  Can you explain more what you mean by "to travel all edges at least once, at minimal cost" and how that relates to "eulerian" which is defined to go across each edge exactly once?

**Comment by claudinepeyrat06** [comment]
I agree, eulerian does not involve edge weights., because you aim to travel each edge exactly once. But a very natural reason to "eulerize" is the so called  chinese postman problem  (also known as route inspection problem)  and its variations,  the postman needs  to travel all edges at least once, starting and ending from the post office. If the graph is eulerian, you look for an eulerian circuit. If it is not eulerian, you need to add edges, but what you want to minimize is the total distance of the new edges.

**Comment by dschult** [comment]
That sounds like it should have a different function, unless there is an easy way to find the edges which eulerize and are minimal. Typically the computation needed for weighted problems is more than for the unweighted problems. 

Wikipedia calls that problem [the Route inspection problem](https://en.wikipedia.org/wiki/Route_inspection_problem) and sketches an algorithm for it.

**Comment by rfulekjames** [comment]
Hi all, I would be happy to fix this small bug (as a first time contributor) in case nobody plans to work on it and possibly also implement the generalization in the direction of [the Route inspection problem](https://en.wikipedia.org/wiki/Route_inspection_problem) with the option _"... to start and end at specified vertices_ ..." in case this is still of interest.

cc @dschult, @claudinepeyrat06

**Comment by rossbar** [comment]
I'm +1 for implementing the fix+tests proposed in the OP. Investigating new algorithms e.g. for route inspection is interesting, but I'd propose to handle that in a separate issue/discussion/PR as that will likely require more discussion. Fixing the defect in `eulerize` is a high priority IMO.

**Comment by dschult** [comment]
Yes -- @rfulekjames please make a PR to fix the bug! We can help with any first-time questions you might have.
Thanks!
